util/encrypt/ase_layer.go

```diff
@@ -0,0 +1,179 @@
+	New: func() interface{} { return make([]byte, blockSize) },
+}
+
+// CtrCipher encrypting data using AES in counter mode
+type CtrCipher struct {
+	nonce uint64
+	block cipher.Block
+}
+
+// NewCtrCipher return a CtrCipher
+func NewCtrCipher(key []byte, nonce uint64) (ctr *CtrCipher, err error) {
+	var block cipher.Block
+	block, err = aes.NewCipher(key)
+	if err != nil {
+		return
+	}
+	ctr = new(CtrCipher)
+	ctr.block = block
+	ctr.nonce = nonce
+	return
+}
+
+// stream returns a cipher.Stream be use to encrypts/decrypts
+func (ctr *CtrCipher) stream(counter uint64) cipher.Stream {
+	counterBuf := make([]byte, aes.BlockSize)
+	binary.BigEndian.PutUint64(counterBuf, ctr.nonce)
+	binary.BigEndian.PutUint64(counterBuf[8:], counter)
+	return cipher.NewCTR(ctr.block, counterBuf)
+}
+
+// Writer implements an io.WriteCloser, it encrypt data using AES before writing to the underlying object.
+type Writer struct {
+	err          error
+	w            io.WriteCloser
+	n            int
+	buf          []byte
+	cipherStream cipher.Stream
+}
+
+// NewWriter returns a new Writer which encrypt data using AES before writing to the underlying object.
+func NewWriter(w io.WriteCloser, ctrCipher *CtrCipher) *Writer {
+	writer := &Writer{w: w}
+	writer.buf = make([]byte, blockSize)
+	writer.cipherStream = ctrCipher.stream(0)
+	return writer
+}
+
+// AvailableSize returns how many bytes are unused in the buffer.
+func (w *Writer) AvailableSize() int { return blockSize - w.n }
+
+// Write implements the io.Writer interface.
+func (w *Writer) Write(p []byte) (n int, err error) {
+	for len(p) > w.AvailableSize() && w.err == nil {
+		copiedNum := copy(w.buf[w.n:], p)
+		w.n += copiedNum
+		err = w.Flush()
+		if err != nil {
+			return
+		}
+		n += copiedNum
+		p = p[copiedNum:]
+	}
+	if w.err != nil {
+		return n, w.err
+	}
+	copiedNum := copy(w.buf[w.n:], p)
+	w.n += copiedNum
+	n += copiedNum
+	return
+}
+
+// Buffered returns the number of bytes that have been written into the current buffer.
+func (w *Writer) Buffered() int { return w.n }
+
+// Flush writes all the buffered data to the underlying object.
+func (w *Writer) Flush() error {
+	if w.err != nil {
+		return w.err
+	}
+	if w.n == 0 {
+		return nil
+	}
+	w.cipherStream.XORKeyStream(w.buf[:w.n], w.buf[:w.n])
+	n, err := w.w.Write(w.buf[:w.n])
+	if n < w.n && err == nil {
+		err = io.ErrShortWrite
+	}
+	if err != nil {
+		w.err = err
+		return err
+	}
+	w.n = 0
+	return nil
+}
+
+// Close implements the io.Closer interface.
+func (w *Writer) Close() (err error) {
+	err = w.Flush()
+	if err != nil {
+		return
+	}
+	return w.w.Close()
+}
+
+// Reader implements an io.ReadAt, reading from the input source after decrypting.
+type Reader struct {
+	r      io.ReaderAt
+	cipher *CtrCipher
+}
+
+// NewReader returns a new Reader which can read from the input source after decrypting.
+func NewReader(r io.ReaderAt, ctrCipher *CtrCipher) *Reader {
+	reader := &Reader{r: r, cipher: ctrCipher}
```

Comment:
Should we define `counterPerBlock` as an attribute of Reader, and evaluate it in NewReader?

Reply:
I think defining `counterPerBlock` as an attribute is better. It makes the reader suitable for the different encrypt block size underlying objects.  How about defining `blockSize ` as an attribute of `CtrCipher`, and as an argument of the `NewCtrCipher`. In this way, I can evaluate the `counterPerBlock` in the `NewReader`.

Reply:
ok